### PR TITLE
(819) Reversal Invoice is submitted to Workday on a submission replaced by a correction

### PR DESCRIPTION
**Pull request by CristinaRO**

### CODE REVIEW

File: app/jobs/submission_reversal_invoice_creation_job.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+class SubmissionReversalInvoiceCreationJob < ApplicationJob
+  def perform(submission)
+    raise "Submission #{submission.id} already has a reversal invoice." if submission.reversal_invoice.present?
+
+    workday_reference = if submission.management_charge.negative?
+                          Workday::SubmitCustomerReversalInvoiceRequest.new(submission).perform
+                        else
+                          Workday::SubmitCustomerReversalInvoiceAdjustmentRequest.new(submission).perform
+                        end
+    submission.create_reversal_invoice!(workday_reference: workday_reference)
```
Comment (tekin) [inline app/jobs/submission_reversal_invoice_creation_job.rb]: if saving this record fails, what happens when the job is re-run? Will Workday let us create a second reversal?

File: app/models/workday/submit_customer_reversal_invoice_adjustment_request.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
+module Workday
+  class SubmitCustomerReversalInvoiceAdjustmentRequest < SubmitCustomerInvoiceAdjustmentRequest
```
[the remark this answers is not in the document]
Comment (tekin) [reply app/models/workday/submit_customer_reversal_invoice_adjustment_request.rb]: I'm wondering if calling this `SubmitCustomerInvoiceReversal` and the other `SubmitCustomerInvoiceAdjustmentReverasal` would be better? It makes it clearer what they're used for, i.e. reversing an invoice or reversing an adjustment. We lose the duplication of the inheritance in the class name, but I don't know how useful that actually is to a reader of the code. What do you think?

File: spec/models/workday/submit_customer_reversal_invoice_adjustment_request_spec.rb [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+    it 'sets Customer_PO_Number with the Submission’s purchase order number' do
+      expect(text_at_xpath('//ns0:Customer_PO_Number')).to eq submission.purchase_order_number
+    end
+
+    it 'sets Memo with the Submission’s ID' do
+      expect(text_at_xpath('//ns0:Memo')).to eq "Submission ID: #{submission.id}"
+    end
+
+    # Question: the original submitter, or the user submitting the correction that triggers this reversal?
```
Comment (tekin) [inline spec/models/workday/submit_customer_reversal_invoice_adjustment_request_spec.rb]: Is this question still a question? if so, how/when will it get answered?

File: spec/models/workday/submit_customer_reversal_invoice_adjustment_request_spec.rb [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    it 'sets Note_Data with the name of the user who submitted the Submission' do
+      expect(text_at_xpath('//ns0:Note_Data//ns0:Note_Content')).to eq 'Forename Surname'
+    end
+
+    it 'sets the invoice as submitted' do
```
Comment (tekin) [inline spec/models/workday/submit_customer_reversal_invoice_adjustment_request_spec.rb]: it looks like this matches the specs for `SubmitCustomerInvoiceAdjustmentRequest`, but I would have expected this to read `'sets the invoice **adjustment** as submitted``

File: spec/models/workday/submit_customer_reversal_invoice_adjustment_request_spec.rb [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    end
+
+    describe '//Customer_Invoice_Line_Replacement_Data' do
+      it 'sets Line_Item_Description with a description of the charge' do
+        expect(
+          text_at_xpath('//ns0:Customer_Invoice_Line_Replacement_Data//ns0:Line_Item_Description')
+        ).to eq 'Reversal of invoice for December 2018 management charge'
+      end
+
+      it 'sets Analytical_Amount as the total spend for the submission' do
```
Comment (tekin) [inline spec/models/workday/submit_customer_reversal_invoice_adjustment_request_spec.rb]: this is actually setting the **negative** of the total spend. Assuming that is what Finance want, this should be reworded to make that clear.

File: spec/models/workday/submit_customer_reversal_invoice_adjustment_request_spec.rb [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        expect(text_at_xpath('//ns0:Customer_Invoice_Line_Replacement_Data//ns0:Analytical_Amount')).to eq '-20.00'
+      end
+
+      it 'sets Extended_Amount as the management charge for the submission' do
+        expect(text_at_xpath('//ns0:Customer_Invoice_Line_Replacement_Data//ns0:Extended_Amount')).to eq '0.20'
```
Comment (tekin) [inline spec/models/workday/submit_customer_reversal_invoice_adjustment_request_spec.rb]: I expected this to be negative! What am I not getting?

File: spec/models/workday/submit_customer_reversal_invoice_request_spec.rb [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    end
+
+    describe '//Customer_Invoice_Line_Replacement_Data' do
+      it 'sets Line_Item_Description with a description of the charge' do
+        expect(
+          text_at_xpath('//ns0:Customer_Invoice_Line_Replacement_Data//ns0:Line_Item_Description')
+        ).to eq 'Reversal of invoice for December 2018 management charge'
+      end
+
+      it 'sets Analytical_Amount as the total spend for the submission' do
```
Comment (tekin) [inline spec/models/workday/submit_customer_reversal_invoice_request_spec.rb]: as above, it's setting the negative of the spend.

File: spec/models/workday/submit_customer_reversal_invoice_request_spec.rb [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        expect(text_at_xpath('//ns0:Customer_Invoice_Line_Replacement_Data//ns0:Analytical_Amount')).to eq '20.00'
+      end
+
+      it 'sets Extended_Amount as the management charge for the submission' do
```
Comment (tekin) [inline spec/models/workday/submit_customer_reversal_invoice_request_spec.rb]: "negative"? "reversal amount"?

File: spec/models/workday/submit_customer_reversal_invoice_adjustment_request_spec.rb (hunk 6 above)
Comment (CristinaRO) [reply spec/models/workday/submit_customer_reversal_invoice_adjustment_request_spec.rb]: "Extended_Amount" is always a positive number. Think of it as the absolute value, whichever direction the money went in.

File: spec/models/submission_spec.rb [hunk 9]
```diff
@@ -27,12 +27,60 @@
...
+      end
+
+      context 'when there is an invoice for the submission' do
+        let!(:invoice) { FactoryBot.create(:submission_invoice, submission: submission) }
+
+        it 'enqueues the creation of a reversal invoice' do
+          allow(SubmissionReversalInvoiceCreationJob).to receive(:perform_later)
+
+          submission.replace_with_no_business
+          expect(SubmissionReversalInvoiceCreationJob).to have_received(:perform_later).with(submission)
```
[the remark this answers is not in the document]
Comment (CristinaRO) [reply spec/models/submission_spec.rb]: 👍

File: spec/models/submission_spec.rb [hunk 10]
```diff
@@ -27,12 +27,60 @@
...
+        end
+      end
+
+      context 'when there is an invoice for the submission' do
+        let!(:invoice) { FactoryBot.create(:submission_invoice, submission: submission) }
+
+        it 'enqueues the creation of a reversal invoice' do
+          allow(SubmissionReversalInvoiceCreationJob).to receive(:perform_later)
+
+          submission.replace_with_no_business
```
[the remark this answers is not in the document]
Comment (CristinaRO) [reply spec/models/submission_spec.rb]: 👍

File: app/jobs/submission_reversal_invoice_creation_job.rb [hunk 11]
```diff
@@ -0,0 +1,12 @@
+class SubmissionReversalInvoiceCreationJob < ApplicationJob
+  def perform(submission)
+    raise "Submission #{submission.id} already has a reversal invoice." if submission.reversal_invoice.present?
```
[the remark this answers is not in the document]
Comment (CristinaRO) [reply app/jobs/submission_reversal_invoice_creation_job.rb]: I don't think this is under normal operations, which is why we want to be alerted if it does occur?

**Comment by CristinaRO** [comment]
Some changes done, see if the naming is clearer now.

Other changes tomorrow.

**Pull request merged**